Copter: Add Airspeed Support

hendjoshsr71 opened the pull request:
To be added after PR https://github.com/ArduPilot/ardupilot/pull/18838 which moves AP_Airspeed usage up into AP_Vehicle

@bnsgeyer 

This adds airspeed support to copter and replaces https://github.com/ArduPilot/ardupilot/pull/8430 & Closes https://github.com/ArduPilot/ardupilot/issues/7654

I tried to incorporate most of the comments in the above PR. It does not incorporate any control elements or autocal. This is reporting only. 

**Notes**: This PR adds 13.6 KB to copter it is defined out for boards 1MB or lower

**Questions not addressed** in above PR: What happens in the EKF if someone tries to use airspeed?

**Tested in SITL only:** 
Type == Analog
![arspd_analog_data_flash](https://user-images.githubusercontent.com/69225461/135593033-746dffdb-9477-4cb9-94ad-f7e3029ab418.png)

Type == MS5525 Weird Data here between mode types
![arspd_ms5525_data_flash](https://user-images.githubusercontent.com/69225461/135593036-9f55ac14-6541-4244-90e6-227e21c31fa3.png)

hendjoshsr71 commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -1476,4 +1486,15 @@ uint8_t GCS_MAVLINK_Copter::high_latency_wind_direction() const
+#if AP_AIRSPEED_ENABLED
+    return GCS_MAVLINK::high_latency_air_temperature();
+#else
+    return INT8_MIN;    // NOTE high latency doesn't specify invalids for fields
```
@stephendade High_latency does not define what fields should be set to in case or invalid data. Any suggestions?

stephendade replied:
I've used 0 to denote any case where to data is unvailable or otherwise invalid in the rest of high_latency, so best to be consistent with that.

hendjoshsr71 replied:
I'm not a fan of 0 for this case. As 0 deg C is certainly possible for air temperature. I'd say either `INT_MAX` or INT_MIN` makes the most sense?

stephendade replied:
Go with ``INT_MIN``. I'll need to raise a PR to fix the rest of HL to make it consistent.

hendjoshsr71 commented:
It would be nice is some the plane & copter version of this could be reconciled and moved up.

**Copter**
https://github.com/ArduPilot/ardupilot/blob/eb9b29f84b940553a2482abca8c1ae3bf2d33542/ArduCopter/GCS_Mavlink.cpp#L196-L204

**Plane**
https://github.com/ArduPilot/ardupilot/blob/eb9b29f84b940553a2482abca8c1ae3bf2d33542/ArduPlane/GCS_Mavlink.cpp#L236-L254

IamPete1 commented:
With this all the vehicles support aispeed (except blimp), I wonder if its worth moving up to AP_Vehicle.

bnsgeyer commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -195,6 +195,16 @@ void GCS_MAVLINK_Copter::send_nav_controller_output() const
+#endif
+
     Vector3f airspeed_vec_bf;
     if (AP::ahrs().airspeed_vector_true(airspeed_vec_bf)) {
```
I just want to point out that the estimated airspeed output by AHRS is a total airspeed meaning it is the length of the airspeed along all three axes.  This is much different than the airspeed provided by the sensor which is along a single axis (x axis).

hendjoshsr71 commented:
I've redone this such that it is built on top of PR https://github.com/ArduPilot/ardupilot/pull/18838 which moves AP_Airspeed usage up into AP_Vehicle

hendjoshsr71 commented:
Rebased this with the latest updates to the moving AP_Airspeed to AP_Vehicle PR.

I've decided to define out this PR for copter and only include it in by default for heli. I think 13.6KB is too high for copter for a feature that will likely be rarely used.

Added a `helicopter.AirspeedDrivers` autotest that is functional and passes. 
Added a SITL change to enable airspeed (and other sensors) for heli

bnsgeyer commented:
@hendjoshsr71, @rmackay9  and @tridge,
I tested the version of this PR before his latest rebase yesterday.   I used my plane since it was set up with an airspeed sensor.  I only had the old Pixhawk airspeed sensors and the other one didn't seem to play well with the newer autopilots with different connectors.  Here is the set up in my plane
3DR Pixhawk 1 (loaded fmuv2 firmware)
3DR GPS
4525D0 airpseed sensor (airspeed_type 1)

I first loaded up the plane version of the PR.  It immediately booted with the airspeed working.  I took some data disarmed and used a leaf blower to provide an airspeed measurement.  [here](https://drive.google.com/file/d/1-1lsz66j0d1nRDvkpsQq_D1m64OycBWj/view?usp=sharing) is a link to the log file.
![image](https://user-images.githubusercontent.com/9907906/140450418-f65cb839-798a-401a-af4e-efb1006db544.png)

Next I compiled and loaded heli firmware.  I had to set the airspeed_type and then reboot.  Once it rebooted, the Device ID populated and it calibrated.  The airspeed calibration looked similar to plane.  Again I collected data disarmed and used the leaf blower to provide airspeed measurement.  [here](https://drive.google.com/file/d/1-4uGxQSw7fEC0_tAZSjLqMpxCOdnbsPt/view?usp=sharing) is a link to the log file.
![image](https://user-images.githubusercontent.com/9907906/140450517-08d1772a-6b63-43b7-b740-2a375b7bb390.png)

hendjoshsr71 commented:
@bnsgeyer Thanks so much for testing!
One note did you see in the console "Airspeed 1 calibrated" I don't see it in the log for heli but do see it in the log for plane. From playing in SITL with the failures messages it seems these messages dont get logged in flash for some reason. 

Rebased this again due to conflicts from the EFI changes.

bnsgeyer commented:
@hendjoshsr71 I will look at the tlog tonight to see if that message came across telemetry.

bnsgeyer commented:
@hendjoshsr71 I looked at the tlogs and there isn't any message for airspeed calibration.  Does it matter when I connect to the vehicle.  I connected shortly after it booted but I wonder if I missed the message or does it have a set sequence of messages it sends when a GCS connects to it.

hendjoshsr71 commented:
You probably just missed it then when connecting. There should be "some" of the following messages,
AirspeedX calibration started
AirspeedX Calibrated 

There are other possible status messages that should have showed up for your failing sensor but didn't. I'm actually a bit surprised status messages (of certain severity) do not get queued for the first connection to be sent out. It is sort of annoying if you miss `AirspeedX Initialization Failed`.

 The real annoyance I have is that they don't end up being logged. Maybe because the logging threads aren't ready for logging by the time the airspeed sensors are initialized?

bnsgeyer commented:
@hendjoshsr71 So where are we at with this?  Do you want me to load up this firmware again and see if I can connect to get the messages?  Or can we push to have this merged since I tested it and it works?

hendjoshsr71 commented:
The missing messages are an annoyance but shouldn't be part of this PR.

I think we should create a follow-up PR to fix the VFR_airspeed issue with body frame x-axis magnitude vs full wind vector magnitude. It is wrong, IMO, but should not cause any detrimental effects and follows what is being don else where. (As it is just reporting info. and we aren't enabling airspeed for copter control here)

So this PR is fine I think after your ground tests. The issue is the preceding PR ([here](https://github.com/pulls)) that moves the Airspeed library up to AP_Vehicle needs testing and issues to be resolved:
-  To be tested on plane
- Check for parameter conversion conflicts due to the zero index of the AP_Airspeed library
    -  I need @tridge's help on this
- Resolve how we want to handle the increase in code size to copter boards that will either push them over the edge of 2MB
    - This MatekH743 PR needs to be merged first, https://github.com/ArduPilot/ardupilot/pull/19195

peterbarker commented on ArduCopter/AP_Arming.cpp:
```diff
@@ -56,14 +56,19 @@ bool AP_Arming_Copter::run_pre_arm_checks(bool display_failure)
+                & alt_checks(display_failure)
+                & AP_Arming::pre_arm_checks(display_failure);
+#if AP_AIRSPEED_ENABLED && !(APM_BUILD_TYPE(APM_BUILD_ArduCopter) && BOARD_FLASH_SIZE < 1025)
+    ret &= AP_Arming::airspeed_checks(display_failure);
```
This doesn't necessarily need to be taen outside the existing structure - could just put it in the middle of the existing list...

hendjoshsr71 replied:
True. Thanks fixed!

Hwurzburg commented:
@tridge @hendjoshsr71 I tested this on a regular plane today....was fine....but I now see that Josh pushed more changes so the testing is probably invalidated....

hendjoshsr71 commented:
Thanks Henry for the testing!!


Those changes were none functional rebase changes required due to moving of
RPM up.

On Wed, Jan 12, 2022, 5:00 PM Henry Wurzburg ***@***.***>
wrote:

> @tridge <https://github.com/tridge> @hendjoshsr71
> <https://github.com/hendjoshsr71> I tested this on a regular plane
> today....was fine....but I now see that Josh pushed more changes so the
> testing is probably invalidated....
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ArduPilot/ardupilot/pull/18813#issuecomment-1011490332>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQQEX5PKTDYXL2E7J2W4LELUVX2ZNANCNFSM5FEGCUIA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bnsgeyer commented:
@hendjoshsr71, @rmackay9  and @tridge 
  I flew this firmware in an actual heli today with the airspeed sensor disabled.  I did some bench tests with the 4525 airspeed sensor and it worked fine.  I couldn't get the DLVR CAN airspeed sensor to work.  Not sure what I am doing wrong.  Hooking it up under I2C DLVR as Airspeed type 9, it recognizes the sensor and shows a device ID but the airspeed reading fluctuates from 0.1 to 0.4.  I also tried connecting as a CAN sensor and set the airspeed type to 8 for that but no luck there either.  In that case it didn't even show a device ID.  So I will need help to check this sensor out.

bnsgeyer commented:
@rmackay9  and @tridge 
Worked with @hendjoshsr71 and was able to get the Matek CAN-DLVR airspeed sensor working with my CUAVv5 controller.

rmackay9 reviewed (approved):
I'm happy with the copter and rover changes..

hendjoshsr71 commented:
See https://github.com/ArduPilot/ardupilot/pull/18838 for the majority of discussion for this PR which was about the moving AP_Airspeed to AP_Vehicle
